inc/elasticsearch_packages/namespace.php

```diff
@@ -0,0 +1,432 @@
+	add_action( 'altis.search.deleted_package', __NAMESPACE__ . '\\on_deleted_package', 10, 4 );
+	add_action( 'altis.search.updated_all_index_settings', __NAMESPACE__ . '\\on_update_index_settings' );
+
+	// Remove default packages updated hook so we don't try to update indexes until
+	// packages have been associated with the domain.
+	add_action( 'plugins_loaded', function () {
+		remove_action( 'altis.search.updated_packages', 'Altis\\Enhanced_Search\\Packages\\on_updated_packages', 10, 2 );
+	}, 20 );
+}
+
+/**
+ * Filter the cron schedules list.
+ *
+ * @param array $schedules The list of available cron schedules.
+ * @return array
+ */
+function cron_schedules( array $schedules ) : array {
+	$schedules['1minute'] = [
+		'display' => __( 'Every minute', 'altis' ),
+		'interval' => 60,
+	];
+	return $schedules;
+}
+
+/**
+ * Set the package upload directory to an S3 directory.
+ *
+ * @param string $path The directory to store packages in.
+ * @return string
+ */
+function packages_dir( string $path ) : string {
+	if ( ! defined( 'S3_UPLOADS_BUCKET' ) || empty( S3_UPLOADS_BUCKET ) ) {
+		return $path;
+	}
+
+	return sprintf( 's3://%s/es-packages', S3_UPLOADS_BUCKET );
+}
+
+/**
+ * Get Elasticsearch Service Client.
+ *
+ * @return ElasticsearchServiceClient
+ */
+function get_es_client() : ElasticsearchServiceClient {
+	return Altis\get_aws_sdk()->createElasticsearchService( [
+		'version' => '2015-01-01',
+	] );
+}
+
+/**
+ * Return the domain ID portion of the Elasticsearch Service host name.
+ */
+function get_elasticsearch_domain() : string {
+	return preg_replace( '#^search-([a-z][a-z0-9\-]+)-[a-z0-9]+\..*$#', '$1', ELASTICSEARCH_HOST );
+}
+
+/**
+ * Create the package on AES and return the package ID string.
+ *
+ * @param string|null $package_id The package file path for ES mappings.
+ * @param string $slug The package slug.
+ * @param string $file The package file path.
+ * @param bool $for_network Whether this is a network level package or not.
+ * @return string|WP_Error
+ */
+function create_package_id( ?string $package_id, string $slug, string $file, bool $for_network = false ) {
+
+	// Check file is an S3 file path.
+	if ( strpos( $file, 's3://' ) !== 0 ) {
+		// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+		return new WP_Error( 'package_file_missing', sprintf( 'The given package filepath %s is not a valid S3 file path.', $file ) );
+	}
+
+	// Check domain.
+	if ( ! defined( 'ELASTICSEARCH_HOST' ) || ! ELASTICSEARCH_HOST ) {
+		return new WP_Error( 'elasticsearch_host_not_found', 'Could not find an AWS ElasticSearch Service Domain to associate the package with.' );
+	}
+
+	// Get AES client.
+	$client = get_es_client();
+
+	// Get Domain ID from host.
+	$domain = get_elasticsearch_domain();
+
+	// Get a unique package name.
+	// Must be no more than 28 chars.
+	$name = substr( sprintf(
+		'p%s-%s',
+		hash( 'crc32', $domain . $file . time() ),
+		trim( basename( $file, '.txt' ), '-' )
```

Comment:
Should this be sanitised?

Reply:
Maybe, but AWS will throw an error if the string is invalid anyway.